doc/port-profile-init/port-profile-init-design.md

```diff
@@ -0,0 +1,179 @@
+
+### Requirements
+
+* SONiC will support both bulk approach and legacy approach.
+* Before creating the ports, SONiC will check if bulk operation is supported by SAI. If not, it will fallback to legacy approach.
+* Ports creation bulk approach in SONiC expectes SAI to not create the port in create_switch() phase.
+* Before creating the ports, even if bulk is suported, SONiC will check if SAI created ports as before. If so, it will fallback to legacy approach.
+* SONiC will send the bulk request only when all ports are configured in APP DB, PORT table. Meaning, only when "PortConfigDone" flag is raised.
+* In case create_ports() API is called with an empty list of ports, SONiC expects SAI to return a status different from SAI_STATUS_NOT_IMPLEMENTED or SAI_STATUS_NOT_SUPPORTED. This call will be made in order to check if bulk approach is supported and implemented in vendor's SAI.
+* After feature is implemented, SAI initialization time should be reduced. Hence, we expect fast-boot up time to meet max 30 seconds restriction.
+
+
+### Architecture Design
+
+This feature does not change the existing SONiC architecture, while it has to change the configuration flow for ports. This change will be covered in orchagent - PortsOrch.
+
+
+### High-Level Design
+
+There will be no changes in CLI, DB schema, YANG model or warm-reboot support.
+
+In order to support feature in SONiC side, Port configuration flow will be changed in sonic-swss PortsOrch module.
+
+<p align=left>
+<img src="img/old_flow.svg" alt="Figure 1. Old Flow">
+</p>
+
+
+<p align=left>
+<img src="img/new_flow.svg" alt="Figure 2. New Flow">
+</p>
+
+
+The above sequence diagram shows that a lot of calls to SAI and SDK APIs are being saved.
+In terms of Fast Reboot, when system is going up, it can save a lot of time and not challenge the 30 seconds restriction.
+
+Notes:
+
+* Multi-Asic is covered for these changes. In Multi-Asic environment, there are multiple instances of swss and Redis DB and each DB has it's own PORT table. Hence, port configuration is per Asic.
+
+* When create_ports() API is called with an empty list of ports, SONiC expects SAI to return any status except of SAI_STATUS_NOT_IMPLEMENTED or SAI_STATUS_NOT_SUPPORTED. This call will be performed in order to validate the API is supported or not.
+
+#### Initialization Flow
+
+As part of PortsOrch initialization we query for all pre-configured ports in SAI and put that in a list for the later comparison logic.
+
+Now that SAI will no longer initialize all ports from SAI profile file, this query is not relevant.
+It will no longer be part of PortsOrch initialization and will only be done in case bulk option is not supported.
+
+2 checks will be performed in initialization flow.
+1. Is bulk option supported? It will be checked using the following flow:
+
```

Comment:
Can we have the explicit switch level capability check for port bulk support instead implicitly assuming with create_ports API? May be we can refer SAI_SWITCH_ATTR_SUPPORTED_FAILOVER_MODE

https://github.com/opencomputeproject/SAI/blob/master/inc/saiswitch.h#L2557

Reply:
Please check the design after changes. 
We allow vendor to implement create_ports() API while not implementing other APIs like remove_ports and set_ports_attribute. 
for this purpose, we will have to check for each API if it is implemented. 

Furthermore, if we will use the suggested approach, it means we will need to wait for PortConfigDone before configuring any port, even if bulk is not supported. and if it is not supported, syncd will translate the bulk command to the regular one. this can delay the process we have today and will not gain the purpose we wanted.